### ISSUE TITLE
Warning for implicitly converting Optional to bool

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#5572]
**Is your feature request related to a problem? Please describe.**

The "truthiness" implemented in `object.__bool__` is a common source of bugs with control flow statements. When combined with `Optional`, it's easy to miss in a code review that `if not x: ...` will take the branch if x is `None` but *also* if `bool(x)` is `False`.

**Describe the solution you’d like**

I propose reporting implicit conversions of `Optional[T]` to `bool`. These include `if x: ...` and `while x: ...` expressions when `x` has the type `Optional[T]`.

There are other cases where implicit conversion to `bool` is done, like `x and ...`, `x or ...`, and possibly more. Reporting these would also be useful.

The warning could suggest changing the expression to `x is not None` or `x is None`. Expressions where explicit conversion is done, like `bool(x)` should not be reported.

**Comment by erictraut** [comment]
Thanks for the suggestion.

We have a principle that we generally don't add new optional checks unless they can be enabled by default in "strict" type checking mode. If a check is not enabled by default in strict mode, it is difficult to discover, and it tends to be used only by the person who requests the feature and a handful of others who happen to stumble across it.

I suspect that this check cannot be enabled by default in strict mode without generating a lot of false positives in existing code bases that use pyright for type checking. I can whip up a prototype to verify my hypothesis — and to determine whether the check finds any legitimate bugs. If my hypothesis is correct, I'm going to be reluctant to implement this without additional signal from other pyright users that it's something they would like to see. If my experiment shows that this check produces very few false positives while uncovering potential legitimate bugs, then I'll be more inclined to implement it and enable it by default in strict mode.

**Comment by kyle-query** [comment]
Thanks for considering it and checking for false positives.

**Comment by kyle-query** [comment]
One downside might be correcting these expressions, when they were written as intended, could mean require changing `if not x` to `if x is None or x = 0`, which is a unsightly and tedious. I don't have a solution to that in mind yet.

**Comment by kyle-query** [comment]
Yes, I agree a linter would be the ideal place to implement this check. However Python linters don't have static type information available. 

There is something sort of similar [here](https://beta.ruff.rs/docs/rules/if-expr-with-true-false/) but I don't think Ruff has type information available, so it could not be configured to only report this when the expression type is Optional[T]. I will look more closely at GitHub's CodeQL, but it seems like it's limited to lexical analysis also.

**Comment by erictraut** [comment]
I prototyped the check. As I hypothesized, the "signal to noise ratio" was very bad. I ran the prototype on a few dozen strictly-typed public repos. New diagnostics were reported for almost all code bases. In some cases there were hundreds of new errors. I examined about a hundred of these cases across several repos, and I didn't see a single actual bug. That means I can't justify enabling such a check as part of "strict" mode. And since we generally don't add checks that can't be enabled by default in "strict" mode, I'm going to decline this feature request. I might reconsider in the future if we receive sufficient signal from pyright users (in the form of upvotes), but I'm going to close the issue for now.
